tests/scripts/all.sh

```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
-    # Test that all the combinations build cleanly. The most common issue is
-    # unused variables/functions, so ensure -Wunused is set.
+build_test_config_combos() {
+    # test that the given file builds with all (valid) combinations of the given options.
+    # syntax: build_test_config_combos FILE VALIDATOR_FUNCTION OPT1 OPT2 ...
+    # The validator function may be "" if all combinations are valid
 
-    msg "build: aes.o for all combinations of relevant config options"
+    FILE=$1
+    shift
+    # this function must echo something iff the clang "-DA -DB ..." string is invalid
+    VALIDATE_OPTIONS=$1
+    shift
+    OPTIONS=("$@")
 
-    for a in set unset; do
-    for b in set unset; do
-    for c in set unset; do
-    for d in set unset; do
-    for e in set unset; do
-    for f in set unset; do
-    for g in set unset; do
-        echo ./scripts/config.py $a MBEDTLS_AES_SETKEY_ENC_ALT
-        echo ./scripts/config.py $b MBEDTLS_AES_DECRYPT_ALT
-        echo ./scripts/config.py $c MBEDTLS_AES_ROM_TABLES
-        echo ./scripts/config.py $d MBEDTLS_AES_ENCRYPT_ALT
-        echo ./scripts/config.py $e MBEDTLS_AES_SETKEY_DEC_ALT
-        echo ./scripts/config.py $f MBEDTLS_AES_FEWER_TABLES
-        echo ./scripts/config.py $g MBEDTLS_PADLOCK_C
-
-        ./scripts/config.py $a MBEDTLS_AES_SETKEY_ENC_ALT
-        ./scripts/config.py $b MBEDTLS_AES_DECRYPT_ALT
-        ./scripts/config.py $c MBEDTLS_AES_ROM_TABLES
-        ./scripts/config.py $d MBEDTLS_AES_ENCRYPT_ALT
-        ./scripts/config.py $e MBEDTLS_AES_SETKEY_DEC_ALT
-        ./scripts/config.py $f MBEDTLS_AES_FEWER_TABLES
-        ./scripts/config.py $g MBEDTLS_PADLOCK_C
-
-        rm -f library/aes.o
-        make -C library aes.o CC="clang" CFLAGS="-O0 -std=c99 -Werror -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wimplicit-fallthrough -Wshadow -Wvla -Wformat=2 -Wno-format-nonliteral -Wshadow -Wasm-operand-widths -Wunused"
-    done
-    done
-    done
-    done
-    done
+    # clear all of the options so that they can be overridden on the clang commandline
+    for OPT in "${OPTIONS[@]}"; do
+        ./scripts/config.py unset ${OPT}
     done
+
+    # enter the directory containing the target file & strip the dir from the filename
+    cd $(dirname ${FILE})
+    FILE=$(basename ${FILE})
+
+    # The most common issue is unused variables/functions, so ensure -Wunused is set.
+    WARNING_FLAGS="-Werror -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wimplicit-fallthrough -Wshadow -Wvla -Wformat=2 -Wno-format-nonliteral -Wshadow -Wasm-operand-widths -Wunused"
+
+    # Extract the command generated by the Makefile to build the target file.
+    # This ensures that we have any include paths, macro definitions, etc
+    # that may be applied by make.
+    # Add -fsyntax-only as we only want a syntax check and don't need to generate a file.
+    COMPILE_CMD=$(make -B -n ${FILE} CC=clang CFLAGS="${WARNING_FLAGS} -fsyntax-only" | egrep "^clang")
```

Comment:
This came about via discussion with @tom-cosgrove-arm , on the basis that it ensures we stay in sync with what Make actually does. I don't have a strong preference, but maybe being slightly fragile is better than potentially silently deviating from make behaviour? Tom WDYT?

Reply:
How about a compromise that leverages the existing makefile? Let it set `LOCAL_CFLAGS` and use that.
```
include library/Makefile

t-%: aes.c; clang -Whard-coded-stuff $(LOCAL_FLAGS) -fsyntax-only aes.c
```

Reply:
Unfortunately including `library/Makefile` breaks (even from a Makefile containing only the line `include library/Makefile`). I couldn't find a simple way to make this work.

Reply:
> maybe being slightly fragile is better than potentially silently deviating from make behaviour?

Yes, that would be my preference, rather than building up a compiler invocation manually

Reply:
> Unfortunately including library/Makefile breaks

Got it: https://github.com/Mbed-TLS/mbedtls/pull/6915#discussion_r1346244686

Reply:
See https://github.com/gilles-peskine-arm/mbedtls/tree/more-aes-checks%2Binclude_Makefile for:

1. A fix for `include library/Makefile` not working.
2. A compromise between the fragile parsing and hard-coding things that should be in the makefile. My current code gets the official `LOCAL_FLAGS`, but doesn't build the dependencies of the normal rule to build the object file. I haven't found a way to do that last thing yet, except with some very akward changes to `library/Makefile`.

(Feel free to ignore the file refactoring, I only did that to make testing easier.)

Reply:
Thanks, have incorporated this into the latest iteration